Fix gRPC bug:   ASSERTION FAILED: initial_metadata_received_

[go/cloud-cxx:grpc-workaround](http://goto.google.com/cloud-cxx:grpc-workaround)

- [ ] Refactor GetRequestMetadataFromContext
- [ ] Add bool for to determine if GetRequestMetadataFromContext gets made
- [x] Handle async rpc case
- [x] Handle streaming rpc case

I tried applying the following patch

```
diff --git a/google/cloud/internal/grpc_opentelemetry.h b/google/cloud/internal/grpc_opentelemetry.h
index 929b6935fb..c64be3154e 100644
--- a/google/cloud/internal/grpc_opentelemetry.h
+++ b/google/cloud/internal/grpc_opentelemetry.h
@@ -18,6 +18,7 @@
 #include "google/cloud/completion_queue.h"
 #include "google/cloud/internal/grpc_metadata_view.h"
 #include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/status_utils.h"
 #include "google/cloud/options.h"
 #include "google/cloud/version.h"
 #include <grpcpp/grpcpp.h>
@@ -86,6 +87,22 @@ T EndSpan(grpc::ClientContext& context, opentelemetry::trace::Span& span,
   return EndSpan(span, std::move(value));
 }
 template <typename T>
 future<T> EndSpan(
     std::shared_ptr<grpc::ClientContext> context,
@@ -94,7 +111,10 @@ future<T> EndSpan(
   return fut.then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
                    c = std::move(context), s = std::move(span)](auto f) {
     auto t = f.get();
-    ExtractAttributes(*c, *s, GrpcMetadataView::kWithServerMetadata);
+    ExtractAttributes(*c, *s,
+                      IsClientOrigin(t.status())
+                          ? GrpcMetadataView::kWithoutServerMetadata
+                          : GrpcMetadataView::kWithServerMetadata);
     DetachOTelContext(oc);
     return EndSpan(*s, std::move(t));
   });

```

Then ran 
```
bazel test google/cloud:internal_grpc_opentelemetry_test
```

and got

```
google/cloud/internal/grpc_opentelemetry_test.cc:149:19:   required from here
./google/cloud/internal/grpc_opentelemetry.h:115:37: error: 'class google::cloud::v2_25::Status' has no member named 'status'; did you mean 'Status'?
  115 |                       IsClientOrigin(t.status())
      |                       ~~~~~~~~~~~~~~^~~~~~~~~~~~
```

Because sometimes the T matches with `Status`, not `StatusOr`.

So I changed the patch to

```
diff --git a/google/cloud/internal/grpc_opentelemetry.h b/google/cloud/internal/grpc_opentelemetry.h
index 929b6935fb..2990cf3083 100644
--- a/google/cloud/internal/grpc_opentelemetry.h
+++ b/google/cloud/internal/grpc_opentelemetry.h
@@ -18,6 +18,7 @@
 #include "google/cloud/completion_queue.h"
 #include "google/cloud/internal/grpc_metadata_view.h"
 #include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/status_utils.h"
 #include "google/cloud/options.h"
 #include "google/cloud/version.h"
 #include <grpcpp/grpcpp.h>
@@ -86,15 +87,34 @@ T EndSpan(grpc::ClientContext& context, opentelemetry::trace::Span& span,
   return EndSpan(span, std::move(value));
 }
 
+future<Status> EndSpan(
+    std::shared_ptr<grpc::ClientContext> context,
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
+    future<Status> fut) {
+  return fut.then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+                   c = std::move(context), s = std::move(span)](auto f) {
+    auto t = f.get();
+    ExtractAttributes(*c, *s,
+                      IsClientOrigin(t)
+                          ? GrpcMetadataView::kWithoutServerMetadata
+                          : GrpcMetadataView::kWithServerMetadata);
+    DetachOTelContext(oc);
+    return EndSpan(*s, std::move(t));
+  });
+}
+
 template <typename T>
-future<T> EndSpan(
+future<StatusOr<T>> EndSpan(
     std::shared_ptr<grpc::ClientContext> context,
     opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
-    future<T> fut) {
+    future<StatusOr<T>> fut) {
   return fut.then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
                    c = std::move(context), s = std::move(span)](auto f) {
     auto t = f.get();
-    ExtractAttributes(*c, *s, GrpcMetadataView::kWithServerMetadata);
+    ExtractAttributes(*c, *s,
+                      IsClientOrigin(t.status())
+                          ? GrpcMetadataView::kWithoutServerMetadata
+                          : GrpcMetadataView::kWithServerMetadata);
     DetachOTelContext(oc);
     return EndSpan(*s, std::move(t));
   });

```
And got 

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
/usr/bin/ld.gold: error: bazel-out/k8-fastbuild/bin/google/cloud/_objs/google_cloud_cpp_grpc_utils_private/grpc_opentelemetry.pic.o: multiple definition of 'google::cloud::v2_25::internal::EndSpan(std::shared_ptr<grpc::ClientContext>, std::shared_ptr<opentelemetry::v1::trace::Span>, google::cloud::v2_25::future<google::cloud::v2_25::Status>)'
/usr/bin/ld.gold: bazel-out/k8-fastbuild/bin/google/cloud/_objs/google_cloud_cpp_grpc_utils_private/async_polling_loop.pic.o: previous definition here
/usr/bin/ld.gold: error: bazel-out/k8-fastbuild/bin/google/cloud/_objs/google_cloud_cpp_grpc_utils_private/minimal_iam_credentials_stub.pic.o: multiple definition of 'google::cloud::v2_25::internal::EndSpan(std::shared_ptr<grpc::ClientContext>, std::shared_ptr<opentelemetry::v1::trace::Span>, google::cloud::v2_25::future<google::cloud::v2_25::Status>)'
/usr/bin/ld.gold: bazel-out/k8-fastbuild/bin/google/cloud/_objs/google_cloud_cpp_grpc_utils_private/async_polling_loop.pic.o: previous definition here
collect2: error: ld returned 1 exit status
```

Can someone who knows more about templates help me here?

--- a/google/cloud/internal/grpc_opentelemetry.cc
+++ b/google/cloud/internal/grpc_opentelemetry.cc
@@ -147,6 +147,24 @@ void ExtractAttributes(grpc::ClientContext& context,
   }
 }
 
+future<Status> EndSpan(
+    std::shared_ptr<grpc::ClientContext> context,
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
+    future<Status> fut) {
+  return fut.then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+                   c = std::move(context), s = std::move(span)](auto f) {
+    auto t = f.get();
+    // If the error is client originated, then do not make the call to get the
+    // gRPC server metadata. Since the call to GetInitialServerMetadata()
+    // *might* crash the program.
+    ExtractAttributes(*c, *s,
+                      IsClientOrigin(t)
+                          ? GrpcMetadataView::kWithoutServerMetadata
+                          : GrpcMetadataView::kWithServerMetadata);
+    DetachOTelContext(oc);
+    return EndSpan(*s, std::move(t));
+  });
+}
 #endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
 
 }  // namespace internal

--- a/google/cloud/internal/grpc_opentelemetry.h
+++ b/google/cloud/internal/grpc_opentelemetry.h
@@ -18,6 +18,7 @@
 #include "google/cloud/completion_queue.h"
 #include "google/cloud/internal/grpc_metadata_view.h"
 #include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/status_utils.h"
 #include "google/cloud/options.h"
 #include "google/cloud/version.h"
 #include <grpcpp/grpcpp.h>
@@ -86,15 +87,26 @@ T EndSpan(grpc::ClientContext& context, opentelemetry::trace::Span& span,
   return EndSpan(span, std::move(value));
 }
 
-template <typename T>
-future<T> EndSpan(
+future<Status> EndSpan(
     std::shared_ptr<grpc::ClientContext> context,
     opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
-    future<T> fut) {
+    future<Status> fut);
+
+template <typename T>
+future<StatusOr<T>> EndSpan(
+    std::shared_ptr<grpc::ClientContext> context,
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
+    future<StatusOr<T>> fut) {
   return fut.then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
                    c = std::move(context), s = std::move(span)](auto f) {
     auto t = f.get();
-    ExtractAttributes(*c, *s, GrpcMetadataView::kWithServerMetadata);
+    // If the error is client originated, then do not make the call to get the
+    // gRPC server metadata. Since the call to GetInitialServerMetadata()
+    // *might* crash the program.
+    ExtractAttributes(*c, *s,
+                      IsClientOrigin(t.status())
+                          ? GrpcMetadataView::kWithoutServerMetadata
+                          : GrpcMetadataView::kWithServerMetadata);
     DetachOTelContext(oc);
     return EndSpan(*s, std::move(t));
   });

--- a/google/cloud/internal/grpc_opentelemetry_test.cc
+++ b/google/cloud/internal/grpc_opentelemetry_test.cc
@@ -51,8 +51,10 @@ using ::google::cloud::testing_util::SpanWithStatus;
 using ::google::cloud::testing_util::StatusIs;
 using ::testing::_;
 using ::testing::AllOf;
+using ::testing::Contains;
 using ::testing::ElementsAre;
 using ::testing::IsEmpty;
+using ::testing::Pair;
 
 TEST(OpenTelemetry, MakeSpanGrpc) {
   namespace sc = ::opentelemetry::trace::SemanticConventions;
@@ -188,6 +190,70 @@ TEST(OpenTelemetry, EndSpanFutureDetachesContext) {
   EXPECT_STATUS_OK(f.get());
   EXPECT_THAT(span_catcher->GetSpans(),
               ElementsAre(SpanNamed("google.cloud.foo.v1.Foo/GetBar")));
+}
+
+class EndSpanGrpcServerMetadataTest : public ::testing::Test {
+ protected:
+  testing_util::ValidateMetadataFixture metadata_fixture_;
+};
+
+TEST_F(EndSpanGrpcServerMetadataTest, UnknownOriginError) {
+  auto span_catcher = InstallSpanCatcher();
+
+  promise<Status> p;
+  auto context = std::make_shared<grpc::ClientContext>();
+  metadata_fixture_.SetServerMetadata(
+      *context, {{{"header", "header-value"}}, {{"trailer", "trailer-value"}}});
+  auto f = EndSpan(context, MakeSpanGrpc("google.cloud.foo.v1.Foo", "GetBar"),
+                   p.get_future());
+  EXPECT_FALSE(f.is_ready());
+  p.set_value(internal::CancelledError("call cancelled"));
+  EXPECT_TRUE(f.is_ready());
+  auto status = f.get();
+  EXPECT_THAT(status, StatusIs(StatusCode::kCancelled));
+
+  EXPECT_THAT(
+      span_catcher->GetSpans(),
+      ElementsAre(AllOf(
+          SpanWithStatus(opentelemetry::trace::StatusCode::kError),
+          SpanHasAttributes(
+              OTelAttribute<std::string>("grpc.peer", "inproc"),
+              OTelAttribute<std::string>("rpc.grpc.response.metadata.header",
+                                         "header-value"),
+              OTelAttribute<std::string>("rpc.grpc.response.metadata.trailer",
+                                         "trailer-value")))));
+}
+
+TEST_F(EndSpanGrpcServerMetadataTest, ClientOrigin) {
+  auto span_catcher = InstallSpanCatcher();
+
+  promise<Status> p;
+  auto context = std::make_shared<grpc::ClientContext>();
+  metadata_fixture_.SetServerMetadata(
+      *context, {{{"header", "header-value"}}, {{"trailer", "trailer-value"}}});
+  auto f = EndSpan(context, MakeSpanGrpc("google.cloud.foo.v1.Foo", "GetBar"),
+                   p.get_future());
+  EXPECT_FALSE(f.is_ready());
+  p.set_value(internal::CancelledError(
+      "call cancelled",
+      GCP_ERROR_INFO().WithMetadata("gl-cpp.error.origin", "client")));
+  EXPECT_TRUE(f.is_ready());
+  auto status = f.get();
+  EXPECT_THAT(status, StatusIs(StatusCode::kCancelled));
+  auto const& metadata = status.error_info().metadata();
+  EXPECT_THAT(metadata, Contains(Pair("gl-cpp.error.origin", "client")));
+
+  EXPECT_THAT(
+      span_catcher->GetSpans(),
+      ElementsAre(AllOf(
+          SpanWithStatus(opentelemetry::trace::StatusCode::kError),
+          SpanHasAttributes(OTelAttribute<std::string>("grpc.peer", "inproc")),
+          Not(SpanHasAttributes(
+              OTelAttribute<std::string>("grpc.peer", "inproc"),
+              OTelAttribute<std::string>("rpc.grpc.response.metadata.header",
+                                         "header-value"),
+              OTelAttribute<std::string>("rpc.grpc.response.metadata.trailer",
+                                         "trailer-value"))))));
 }
 
 TEST(OpenTelemetry, TracedAsyncBackoffEnabled) {

